tests/integration/expected/and_u32.masm

```diff
@@ -0,0 +1,611 @@
+          swap.1
+          push.32
+          movup.3
+          u32.sub.checked
+          u32.shr.checked
+          u32.or
+        else
+          dup.0
+          u32.add.checked.1
+          padw
+          movup.4
+          mem_loadw
+          movdn.4
+          drop
+          drop
+          drop
+          push.32
+          dup.3
+          u32.sub.checked
+          u32.shr.checked
+          swap.1
+          padw
+          movup.4
+          mem_loadw
+          drop
+          drop
+          drop
+          movup.2
+          u32.shl.checked
+          u32.or
+        end
+      end
+    end
+  end
+
+end
+
+export.realign_dw
+  dup.3
+  u32.shl.checked
+  movdn.2
+  dup.0
+  push.32
+  dup.4
+  u32.shr.checked
+  movup.4
+  u32.or
+  movdn.2
+  dup.3
+  u32.shl.checked
+  swap.1
+  push.32
+  movup.4
+  u32.shr.checked
+  u32.or
+  swap.1
+
+end
+
+export.load_dw
+  dup.2
+  eq.0
+  dup.3
+  push.8
+  u32.lt.checked
+  assert
+  if.true
+    movup.2
+    drop
+    dup.1
+    eq.0
+    if.true
+      swap.1
+      drop
+      padw
+      movup.4
+      mem_loadw
+      movup.4
+      movup.4
+      drop
+      drop
+    else
+      dup.1
+      eq.1
+      if.true
+        swap.1
+        drop
+        padw
+        movup.4
+        mem_loadw
+        drop
+        movup.3
+        drop
+      else
+        swap.1
+        eq.2
+        if.true
+          padw
+          movup.4
+          mem_loadw
+          drop
+          drop
+        else
+          dup.0
+          u32.add.checked.1
+          padw
+          movup.4
+          mem_loadw
+          movup.4
+          movup.4
+          movup.4
+          drop
+          drop
+          drop
+          movup.4
+          padw
+          movup.4
+          mem_loadw
+          drop
+          drop
+          drop
+        end
+      end
+    end
+  else
+    dup.1
+    eq.0
+    if.true
+      swap.1
+      drop
+      padw
+      movup.4
+      mem_loadw
+      movup.4
+      drop
+      exec.realign_dw
+    else
+      dup.1
+      eq.1
+      if.true
+        swap.1
+        drop
+        padw
+        movup.4
+        mem_loadw
+        drop
+        exec.realign_dw
+      else
+        swap.1
+        eq.2
+        if.true
+          dup.0
+          u32.add.checked.1
+          padw
+          movup.4
+          mem_loadw
+          movup.4
+          movup.4
+          movup.4
+          drop
+          drop
+          drop
+          swap.1
+          padw
+          movup.4
+          mem_loadw
+          drop
+          drop
+          exec.realign_dw
+        else
+          dup.0
+          u32.add.checked.1
+          padw
+          movup.4
+          mem_loadw
+          movup.4
+          movup.4
+          drop
+          drop
+          movup.2
+          padw
+          movup.4
+          mem_loadw
+          drop
+          drop
+          drop
+          exec.realign_dw
+        end
+      end
+    end
+  end
+
+end
+
+export.entrypoint
+  u32.and
```

Comment:
Since i32 and mem intrinsics modules are always linked by `MasmCompiler::compile` we end up with a lot of dead code. Shouldn't we link them only when those functions are used?

Reply:
It's a bit tricky, because they aren't called in a normal sense like regular functions, they are the result of lowering instructions in the IR to MASM. In order to detect their use, we have to check if any of the instructions backed by intrinsics are used. We can certainly be cleverer about this in the future, but in the short term it seemed likely that we would need both intrinsics modules in practice, so felt like a low priority to me. In the near future though, mainly when emitting programs, I would like to do a proper dead code elimination pass to strip out any unused/dead code.

Reply:
Makes sense. For some reason, I thought we would not want to run transformation passes on MASM, but yeah, DCE pass on MASM might be easier than trying to avoid generating a dead code.

Reply:
It's easier to run transformations on the IR, but due to things like this, it becomes necessary sometimes to run transformations on the emitted MASM as well. The peephole optimization pass I've mentioned before is another example - since lowering from the IR isn't 100% perfect, we can further optimize the emitted MASM to improve the output of codegen without changing the semantics of the program. We are necessarily limited in what kinds of transformations we can do safely, but there is still plenty of room for optimization even at the MASM level.